### ISSUE TITLE
How to train a model with patch_size = 2?

### ISSUE THREAD

**Issue by iamweiweishi** [madebyollin/taehv#22]
Hi
I noticed that there is a parameter patch_size in taehv.py, which could save much memory for high resolution.
But when I trained with patch_size=2, the results are very blurry. 
May I know how to improve the reconstruction results? Is there any tricks on the issue?

**Comment by madebyollin** [comment]
My patch_size>1 models are kinda blurry too 😭 but I train for a long time and also use GAN (adversarial loss) to encourage generation of fake (non-reconstructed) sharp detail. You can see more discussion of training TAEW2.2 (one of the patch_size=2 models) here https://github.com/madebyollin/taehv/issues/17 and see examples of the fake detail in https://huggingface.co/datasets/madebyollin/movirec, e.g. by comparing

**Source Frames**

![](https://huggingface.co/datasets/madebyollin/movirec/resolve/main/video-patch-results/source/stills/batch_17_1evdbx0_1evgklr_1evrdj8_1evsffd_1ewa5jf.png)


... to

**TAEW2.2 Reconstructed Frames**


![](https://huggingface.co/datasets/madebyollin/movirec/resolve/main/video-patch-results/vae-TAEW2.2/stills/batch_17_1evdbx0_1evgklr_1evrdj8_1evsffd_1ewa5jf.png)

...you can see that a lot of the sharp texture detail in the TAEW2.2 output is fake (generated) rather than reconstructed.

**Comment by c-zt-p** [comment]
Hi,

Thank you for your great work on TAEHV. I'm currently training the WAN2.2 VAE and encountering the same issue of blurry reconstruction. I'm trying to add a GAN loss as you said. However, I'm facing difficulties in choosing the right weight. When I set the GAN loss weight relative to the MSE loss as 1:1, the GAN loss quickly drops to nearly zero and seems to have no effect on visual quality. When I increase the weight to, say, 100, the GAN loss does work but severely degrades the reconstruction PSNR and overall fidelity.

I was wondering what strategy you use for setting the weight of the GAN loss vs. the reconstruction losses (MSE and LPIPS) when training the TAEHV VAE? How did you balance them to achieve both sharp details and stable reconstruction performance?

Any insights and advice would be greatly appreciated. Thank you again!

**Comment by madebyollin** [comment]
@c-zt-p A few things:
1. I typically [rescale](https://github.com/madebyollin/seraena/blob/main/seraena.py#L141) the discriminator gradients/corrections before using them, so the adv loss weight directly controls the gradient magnitude. From there it's a question of making the adv gradient magnitude higher-than-but-not-overwhelmingly-higher-than the reconstruction gradient magnitudes. iirc I typically ended up multiplying LPIPS by 0.1 and GAN by 10 to get the gradients scaled reasonably against 1x MSE, but you'd want to actually look at your gradients (from each loss term as well as combined) to make sure they're reasonable (since there's lots of ways for your gradients to end up broken).
2. I typically do scheduled adv weight warmup, so the decoder is pretrained with reconstruction losses only, then gradually gets pushed to the "closest-to-good-reconstruction" data mode as the adversarial loss warms up to maximum.
3. Rather than having a single MSE loss, I usually apply MSE/MAE loss at multiple pooled resolutions (e.g. full-res and 1/8 res) with higher weight for the lower-resolution reconstruction losses, so that the GAN loss is only dominating for high frequencies / fine details.

**Comment by c-zt-p** [comment]
Thank you so much! I followed your suggestions (GAN warmup, multi-resolution MSE) and trained for 200k steps now. My results on DAVIS:

PSNR: 26.1, SSIM: 0.75 (close to TAEHV)
LPIPS: 0.19, FVD: 580 (TAEHV: ~0.14, ~380, which are far better than mine)

Another issue is that when I decode latent from WAN2.2's DiT using my VAE, the output has a noticeable color misalignment compared to TAEHV. Do you have any experience or strategy aligning latent statistics or adjusting decoder for downstream compatibility?

WAN2.2-5B DiT + WAN2.2 VAE:
<img width="1287" height="447" alt="Image" src="https://github.com/user-attachments/assets/d1e1dc55-8fea-475a-aa4c-3280fd683002" />
WAN2.2-5B DiT + TAEHV_WAN2.2:
<img width="1291" height="453" alt="Image" src="https://github.com/user-attachments/assets/64832c81-40ef-4d06-aa92-a692eb60f50e" />
WAN2.2-5B DiT + mine
<img width="1287" height="452" alt="Image" src="https://github.com/user-attachments/assets/5226f68e-fa8d-41a0-80c9-0adc5afa7d60" />

I’ll share more details if needed. Thank you!

**Comment by madebyollin** [comment]
> Another issue is that when I decode latent from WAN2.2's DiT using my VAE, the output has a noticeable color misalignment compared to TAEHV

I'd recommend printing out the RMS of the latents you're training on vs testing on. The HuggingFace VAE implementations always require you to apply a VAE transformation outside of the VAE module in order to make latents ~normalized, e.g. the two `IMPORTANT` lines below:

```python3
#!/usr/bin/env python3
import torch
import torch.nn as nn
from diffusers import AutoencoderKLWan

DEVICE = "cuda"
DTYPE = torch.bfloat16
IMAGE_CHANNELS = 3

class RefVAE(nn.Module):
    def __init__(self):
        super().__init__()
        self.vae = AutoencoderKLWan.from_pretrained(
            "Wan-AI/Wan2.2-TI2V-5B-Diffusers", subfolder="vae", torch_dtype=DTYPE
        )
        self.latent_channels = self.vae.config.z_dim
        shape = (1, self.latent_channels, 1, 1, 1)
        kw = dict(device=DEVICE, dtype=DTYPE)
        self.latents_mean = torch.tensor(self.vae.config.latents_mean).view(shape).to(**kw)
        # ???? the huggingface code inverts this for some reason...
        self.latents_std = torch.tensor(self.vae.config.latents_std).view(shape).to(**kw).reciprocal_()

    @torch.no_grad()
    def encode_seq(self, x):
        assert x.ndim == 5 and x.shape[1] == IMAGE_CHANNELS  # NCTHW
        in_dtype = x.dtype
        y = x.to(DTYPE).mul(2).sub_(1)
        y = self.vae.encode(y).latent_dist.sample()
        y = y.sub_(self.latents_mean).mul_(self.latents_std)  # IMPORTANT: apply VAE transformation manually
        return y.to(in_dtype)

    @torch.no_grad()
    def decode_seq(self, x):
        assert x.ndim == 5 and x.shape[1] == self.latent_channels  # NCTHW
        in_dtype = x.dtype
        y = x.to(DTYPE).div(self.latents_std).add_(self.latents_mean) # IMPORTANT: apply VAE transformation manually
        y = self.vae.decode(y).sample
        y = y.mul_(0.5).add_(0.5).clamp_(0, 1)
        return y.to(in_dtype)
```

The DiT is trained to generate normalized/transformed latents, so if the distilled VAE isn't trained on normalized/transformed latents, there will be a mismatch.

**Comment by c-zt-p** [comment]
Fixing transformation really works! Now I can generate correct video outputs with my model. A lot of thanks!

<img width="1267" height="641" alt="Image" src="https://github.com/user-attachments/assets/9f20f189-c008-449e-beab-28296459764d" />

**Comment by c-zt-p** [comment]
Another small question. Is there any reason for you to choose MSE as the alignment loss for TAEHV latent training instead of L1 or some other additional losses? Just curious.

**Comment by madebyollin** [comment]
> Another small question. Is there any reason for you to choose MSE as the alignment loss for TAEHV latent training instead of L1 or some other additional losses? Just curious.

@c-zt-p No, not particularly; L1 would probably work the same or better (i.e. slightly sharper). I usually rely on an additional [decoder feature-matching loss term](https://github.com/madebyollin/taehv/issues/17#issuecomment-4359593832) if I need to make encoded latents "sharp" / in-distribution for the decoder, so I haven't paid that much attention to L1 vs L2 behavior.
